File: book/snippets/js/src/messages/filters.ts
import { Filter, Keys, Kind, EventBuilder, Timestamp, Tag } from "@rust-nostr/nostr-sdk";

export async function run() {
    // Generate keys and Events
    const keys = Keys.generate();
    const keys2 = Keys.generate();

    const kind0 = new Kind(0);
    const kind1 = new Kind(1);
    const kind4 = new Kind(4);

    const event = EventBuilder.textNote("Hello World!").signWithKeys(keys);
    const event2 = new EventBuilder(kind0, "Goodbye World!")
        .tags([Tag.identifier("Identification D Tag")])
        .signWithKeys(keys2);

    console.log();
    console.log("Creating Filters:");

    // ANCHOR: create-filter-id
    // Filter for specific ID
    console.log("  Filter for specific Event ID:");
    let f = new Filter().id(event.id);
    console.log(`     ${f.asJson()}`);
    // ANCHOR_END: create-filter-id

    console.log();
    // ANCHOR: create-filter-author
    // Filter for specific Author
    console.log("  Filter for specific Author:");
    f = new Filter().author(keys.publicKey);
    console.log(`     ${f.asJson()}`);
    // ANCHOR_END: create-filter-author

    console.log();
    // ANCHOR: create-filter-kind-pk
    // Filter by PK and Kinds
    console.log("  Filter with PK and Kinds:");
    f = new Filter()
        .pubkey(keys.publicKey)
        .kind(kind1);
    console.log(`     ${f.asJson()}`);
    // ANCHOR_END: create-filter-kind-pk

    console.log();
    // ANCHOR: create-filter-search
    // Filter for specific string
    console.log("  Filter for specific search string:");
    f = new Filter().search("Ask Nostr Anything");
    console.log(`     ${f.asJson()}`);
    // ANCHOR_END: create-filter-search

    console.log();
    // ANCHOR: create-filter-timeframe
    console.log("  Filter for events from specific public key within given timeframe:");
    // Create timestamps
    const date = new Date(2009, 1, 3, 0, 0);
    const timestamp = Math.floor(date.getTime() / 1000);
    const sinceTs = Timestamp.fromSecs(timestamp);
    const untilTs = Timestamp.now();

    // Filter with timeframe
    f = new Filter()
        .pubkey(keys.publicKey)
        .since(sinceTs)
        .until(untilTs);
    console.log(`     ${f.asJson()}`);
    // ANCHOR_END: create-filter-timeframe

    console.log();
    // ANCHOR: create-filter-limit
    // Filter for specific PK with limit
    console.log("  Filter for specific Author, limited to 10 Events:");
    f = new Filter()
        .author(keys.publicKey)
        .limit(10);
    console.log(`     ${f.asJson()}`);
    // ANCHOR_END: create-filter-limit

    console.log();
    // ANCHOR: create-filter-hashtag
    // Filter for Hashtags
    console.log("  Filter for a list of Hashtags:");
    f = new Filter().hashtags(["#Bitcoin", "#AskNostr", "#Meme"]);
    console.log(`     ${f.asJson()}`);
    // ANCHOR_END: create-filter-hashtag

    console.log();
    // ANCHOR: create-filter-reference
    // Filter for Reference
    console.log("  Filter for a Reference:");
    f = new Filter().reference("This is my NIP-12 Reference");
    console.log(`     ${f.asJson()}`);
    // ANCHOR_END: create-filter-reference

    console.log();
    // ANCHOR: create-filter-identifier
    // Filter for Identifier
    console.log("  Filter for a Identifier:");
    f = new Filter().identifier("This is my NIP-12 Identifier");
    console.log(`     ${f.asJson()}`);
    // ANCHOR_END: create-filter-identifier

    console.log();
    console.log("Modifying Filters:");
    // ANCHOR: modify-filter
    // Modifying Filters (adding/removing)
    f = new Filter()
        .pubkeys([keys.publicKey, keys2.publicKey])
        .ids([event.id, event2.id])
        .kinds([kind0, kind1])
        .author(keys.publicKey);

    // Add an additional Kind to existing filter
    f = f.kinds([kind4]);

    // Print Results
    console.log("  Before:");
    console.log(`     ${f.asJson()}`);
    console.log();

    // Remove PKs, Kinds and IDs from filter
    f = f.removePubkeys([keys2.publicKey]);
    console.log(" After (remove pubkeys):");
    console.log(`     ${f.asJson()}`);
    const kind_rem0 = new Kind(0);
    const kind_rem4 = new Kind(4);
    f = f.removeKinds([kind_rem0, kind_rem4]);
    console.log("  After (remove kinds):");
    console.log(`     ${f.asJson()}`);

    f = f.removeIds([event2.id]);
    console.log("  After (remove IDs):");
    console.log(`     ${f.asJson()}`);
    // ANCHOR_END: modify-filter

    console.log();
    console.log("Other Filter Operations:");
    // ANCHOR: other-parse
    // Parse filter
    console.log("  Parse Filter from Json:");
    const fJson = f.asJson();
    f = Filter.fromJson(fJson);
    console.log(`     ${f.asJson()}`);
    // ANCHOR_END: other-parse

    console.log();
    // ANCHOR: other-match
    console.log("  Logical tests:");
    const kind_match = new Kind(1);
    f = new Filter().author(keys.publicKey).kind(kind_match);
    console.log(`     Event match for filter: ${f.matchEvent(event)}`);
    console.log(`     Event2 match for filter: ${f.matchEvent(event2)}`);
    // ANCHOR_END: other-match
}

File: book/snippets/python/src/messages/filters.py
from nostr_sdk import Filter, FilterRecord, Keys, Kind, EventBuilder, Timestamp, Tag
import time, datetime


def filters():
    # Generate keys and Events
    keys = Keys.generate()
    keys2 = Keys.generate()
    event = EventBuilder.text_note("Hello World!").sign_with_keys(keys)
    event2 = EventBuilder(Kind(1), "Goodbye World!").tags([Tag.identifier("Identification D Tag")]).sign_with_keys(keys2)

    print()
    print("Creating Filters:")

    # ANCHOR: create-filter-id
    # Filter for specific ID
    print("  Filter for specific Event ID:")
    f = Filter().id(event.id())
    print(f"     {f.as_json()}")
    # ANCHOR_END: create-filter-id

    print()
    # ANCHOR: create-filter-author
    # Filter for specific Author
    print("  Filter for specific Author:")
    f = Filter().author(keys.public_key())
    print(f"     {f.as_json()}")
    # ANCHOR_END: create-filter-author

    print()
    # ANCHOR: create-filter-kind-pk
    # Filter by PK and Kinds
    print("  Filter with PK and Kinds:")
    f = Filter()\
        .pubkey(keys.public_key())\
        .kind(Kind(1))
    print(f"     {f.as_json()}")
    # ANCHOR_END: create-filter-kind-pk

    print()
    # ANCHOR: create-filter-search
    # Filter for specific string
    print("  Filter for specific search string:")
    f = Filter().search("Ask Nostr Anything")
    print(f"     {f.as_json()}")
    # ANCHOR_END: create-filter-search

    print()
    # ANCHOR: create-filter-timeframe
    print("  Filter for events from specific public key within given timeframe:")
    # Create timestamps
    date = datetime.datetime(2009, 1, 3, 0, 0)
    timestamp = int(time.mktime(date.timetuple()))
    since_ts = Timestamp.from_secs(timestamp)
    until_ts = Timestamp.now()

    # Filter with timeframe
    f = Filter()\
        .pubkey(keys.public_key())\
        .since(since_ts)\
        .until(until_ts)
    print(f"     {f.as_json()}")
    # ANCHOR_END: create-filter-timeframe

    print()
    # ANCHOR: create-filter-limit
    # Filter for specific PK with limit
    print("  Filter for specific Author, limited to 10 Events:")
    f = Filter()\
        .author(keys.public_key())\
        .limit(10)
    print(f"     {f.as_json()}")
    # ANCHOR_END: create-filter-limit

    print()
    # ANCHOR: create-filter-hashtag
    # Filter for Hashtags
    print("  Filter for a list of Hashtags:")
    f = Filter().hashtags(["#Bitcoin", "#AskNostr", "#Meme"])
    print(f"     {f.as_json()}")
    # ANCHOR_END: create-filter-hashtag

    print()
    # ANCHOR: create-filter-reference
    # Filter for Reference
    print("  Filter for a Reference:")
    f = Filter().reference("This is my NIP-12 Reference")
    print(f"     {f.as_json()}")
    # ANCHOR_END: create-filter-reference

    print()
    # ANCHOR: create-filter-identifier
    # Filter for Identifier
    print("  Filter for a Identifier:")
    identifier = event2.tags().identifier()
    if identifier is not None:
        f = Filter().identifier(identifier)
        print(f"     {f.as_json()}")
    # ANCHOR_END: create-filter-identifier

    print()
    print("Modifying Filters:")
    # ANCHOR: modify-filter
    # Modifying Filters (adding/removing)
    f = Filter()\
        .pubkeys([keys.public_key(), keys2.public_key()])\
        .ids([event.id(), event2.id()])\
        .kinds([Kind(0), Kind(1)])\
        .author(keys.public_key())

    # Add an additional Kind to existing filter
    f = f.kinds([Kind(4)])

    # Print Results
    print("  Before:")
    print(f"     {f.as_json()}")
    print()

    # Remove PKs, Kinds and IDs from filter
    f = f.remove_pubkeys([keys2.public_key()])
    print(" After (remove pubkeys):")
    print(f"     {f.as_json()}")

    f = f.remove_kinds([Kind(0), Kind(4)])
    print("  After (remove kinds):")
    print(f"     {f.as_json()}")

    f = f.remove_ids([event2.id()])
    print("  After (remove IDs):")
    print(f"     {f.as_json()}")
    # ANCHOR_END: modify-filter

    print()
    print("Other Filter Operations:")
    # ANCHOR: other-parse
    # Parse filter
    print("  Parse Filter from Json:")
    f_json = f.as_json()
    f = Filter().from_json(f_json)
    print(f"     {f.as_record()}")
    # ANCHOR_END: other-parse

    print()
    # ANCHOR: other-record
    print("  Construct Filter Record and extract author:")
    # Filter Record
    fr = FilterRecord(ids=[event.id()],authors=[keys.public_key()], kinds=[Kind(0)], search="", since=None, until=None, limit=1, generic_tags=[])
    f = Filter().from_record(fr)
    print(f"     {f.as_json()}")
    # ANCHOR_END: other-record

    print()
    # ANCHOR: other-match
    print("  Logical tests:")
    f = Filter().author(keys.public_key()).kind(Kind(1))
    print(f"     Event match for filter: {f.match_event(event)}")
    print(f"     Event2 match for filter: {f.match_event(event2)}")
    # ANCHOR_END: other-match

if __name__ == '__main__':
   filters()